File: cpp/samples/transport/sensorsdataflow.cpp
/**
 * @file sensorsdataflow.cpp
 * @brief Demonstrates usage of vista-sdk for transforming proprietary sensor data to ISO19848
 * @details This sample shows how to use DataChannelList as a mapping layer between
 *          proprietary system IDs (short_id) and ISO19848 LocalIds, then aggregate sensor
 *          readings into TabularData groups based on updateCycle (ISO19848 optimization)
 */

#include <dnv/vista/sdk/core/VIS.h>
#include <dnv/vista/sdk/transport/serialization/json/datachannel/Extensions.h>
#include <dnv/vista/sdk/transport/serialization/json/timeseries/Extensions.h>

#include <algorithm>
#include <iomanip>
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <vector>

int main()
{
    using namespace dnv::vista::sdk;
    using namespace dnv::vista::sdk::transport;
    namespace jdc = dnv::vista::sdk::transport::serialization::json::datachannel;
    namespace jts = dnv::vista::sdk::transport::serialization::json::timeseries;

    struct SensorReading
    {
        std::string systemId;
        double value;
        DateTimeOffset timestamp;
        std::string quality;
    };

    auto formatDouble = [](double value) {
        std::string str = std::to_string(value);
        str.erase(str.find_last_not_of('0') + 1, std::string::npos);
        if (str.back() == '.')
        {
            str.pop_back();
        }
        return str;
    };

    auto createDataChannelList = []() {
        datachannel::DataChannelList dcList;

        auto localId1Opt =
            LocalId::fromString("/dnv-v2/vis-3-4a/411.1-1/C101.63/S206/meta/qty-temperature/cnt-cooling.air");
        if (!localId1Opt.has_value())
        {
            throw std::runtime_error{ "Failed to parse localId1" };
        }
        auto localId1 = localId1Opt.value();

        auto dcId1 = datachannel::DataChannelId{ localId1, std::optional<std::string>{ "TEMP001" }, std::nullopt };

        auto restriction1 = datachannel::Restriction{};
        restriction1.setFractionDigits(1);
        restriction1.setMaxInclusive(200.0);
        restriction1.setMinInclusive(-50.0);

        auto format1 = datachannel::Format{ "Decimal", restriction1 };
        auto dataChannelType1 = datachannel::DataChannelType{ "Inst", 1.0 };
        auto range1 = datachannel::Range{ 0.0, 150.0 };
        auto unit1 = datachannel::Unit{ "°C", "Temperature" };

        auto prop1 = datachannel::Property{
            dataChannelType1, format1, range1, unit1, "OPC_QUALITY", std::nullopt, "Main Engine Air Cooler Temperature"
        };
        dcList.add(datachannel::DataChannel{ dcId1, prop1 });

        auto localId2Opt = LocalId::fromString("/dnv-v2/vis-3-4a/511.15-1/E32/meta/qty-power");
        if (!localId2Opt.has_value())
        {
            throw std::runtime_error{ "Failed to parse localId2" };
        }
        auto localId2 = localId2Opt.value();

        auto dcId2 = datachannel::DataChannelId{ localId2, std::optional<std::string>{ "PWR001" }, std::nullopt };
        auto format2 = datachannel::Format{ "Decimal", datachannel::Restriction{} };
        auto dataChannelType2 = datachannel::DataChannelType{ "Inst", 1.0 };
        auto range2 = datachannel::Range{ 0.0, 10000.0 };
        auto unit2 = datachannel::Unit{ "kW", "Power" };

        auto prop2 = datachannel::Property{ dataChannelType2,        format2, range2, unit2, std::nullopt, std::nullopt,
                                            "Generator Power Output" };
        dcList.add(datachannel::DataChannel{ dcId2, prop2 });

        auto localId3Opt = LocalId::fromString("/dnv-v2/vis-3-4a/621.22i/S110/meta/cnt-diesel.oil/cmd-stop");
        if (!localId3Opt.has_value())
        {
            throw std::runtime_error{ "Failed to parse localId3" };
        }
        auto localId3 = localId3Opt.value();

        auto dcId3 = datachannel::DataChannelId{ localId3, std::optional<std::string>{ "ALT001" }, std::nullopt };
        auto format3 = datachannel::Format{ "Boolean", datachannel::Restriction{} };
        auto dataChannelType3 = datachannel::DataChannelType{ "Command" };

        auto prop3 = datachannel::Property{
            dataChannelType3, format3, std::nullopt, std::nullopt, std::nullopt, std::nullopt, "Diesel Oil Stop Command"
        };
        dcList.add(datachannel::DataChannel{ dcId3, prop3 });

        auto shipIdOpt = ShipId::fromString("IMO1234567");
        if (!shipIdOpt.has_value())
        {
            throw std::runtime_error{ "Failed to parse shipId" };
        }
        auto shipId = shipIdOpt.value();
        auto configRef = datachannel::ConfigurationReference{ "DataChannelList-v1",
                                                              DateTimeOffset{ "2026-07-17T00:00:00Z" },
                                                              "1.0" };
        auto versionInfo = datachannel::VersionInformation{ "dnv", "v2", "https://docs.vista.dnv.com" };
        auto header = datachannel::Header{
            shipId, configRef, versionInfo, "Vista SDK Sample", DateTimeOffset{ "2026-07-17T00:00:00Z" }
        };
        auto package = datachannel::Package{ header, dcList };
        return datachannel::DataChannelListPackage{ package };
    };

    std::cout << "=== vista-sdk SensorDataFlow Sample ===\n\n";

    {
        std::cout << "1. DataChannelList: Creating sample data channel list\n";
        std::cout << "--------------------------------------------------------\n";

        auto dclPackage = createDataChannelList();

        std::cout << "Created DataChannelList with " << dclPackage.dataChannelList().size() << " channels:\n";

        for (const auto& dc : dclPackage.dataChannelList())
        {
            auto cycle = dc.property().dataChannelType().updateCycle();
            std::string cycleStr = cycle.has_value() ? formatDouble(cycle.value()) + "s" : "None";
            std::cout << "  " << dc.dataChannelId().shortId().value_or("<no-short-id>") << " (updateCycle=" << cycleStr
                      << ")\n";
        }

        {
            std::cout << "\n";
            auto jsonStr = jdc::toJsonString(dclPackage, true);
            std::cout << "Serialized to JSON: " << jsonStr.size() << " bytes\n";
            std::cout << "Package:\n" << jsonStr << "\n";
        }
        std::cout << "\n";

        std::cout << "2. Proprietary sensor data: Simulating sensor readings\n";
        std::cout << "---------------------------------------------------------\n";

        auto baseTime = DateTimeOffset{ 2026, 7, 17, 10, 0, 0, TimeSpan{} };

        std::vector<SensorReading> readings{
            { "TEMP001", 45.2, baseTime, "Good" },  { "TEMP001", 46.1, baseTime + TimeSpan::fromMinutes(5), "Good" },
            { "PWR001", 2500.0, baseTime, "Good" }, { "PWR001", 2650.5, baseTime + TimeSpan::fromMinutes(5), "Good" },
            { "ALT001", 1.0, baseTime, "Good" },    { "UNKNOWN", 123.4, baseTime, "Good" },
        };

        std::cout << "Created " << readings.size() << " sensor readings from proprietary system\n";
        std::cout << "Sample readings:\n";
        for (size_t i = 0; i < readings.size(); ++i)
        {
            std::cout << "  SystemID=" << readings[i].systemId << ", Value=" << readings[i].value
                      << ", Timestamp=" << readings[i].timestamp.toString() << "\n";
        }
        std::cout << "\n";

        std::cout << "3. Validation: Filter readings against DataChannelList\n";
        std::cout << "--------------------------------------------------------\n";

        std::map<std::string, std::vector<SensorReading>> validReadings;

        for (const auto& reading : readings)
        {
            auto dcOpt = dclPackage.dataChannelList().from(reading.systemId);
            if (dcOpt.has_value())
            {
                validReadings[reading.systemId].push_back(reading);
                std::cout << "  Accepted: " << reading.systemId << " = " << reading.value << "\n";
            }
            else
            {
                std::cout << "  Rejected: " << reading.systemId << " (unknown channel)\n";
            }
        }

        std::cout << "\nAccepted " << validReadings.size() << " unique channels with valid data\n";
        std::cout << "\n";

        std::cout << "4. Grouping: Group channels by updateCycle (ISO19848 optimization)\n";
        std::cout << "---------------------------------------------------------------------\n";

        std::map<std::string, std::vector<std::pair<datachannel::DataChannel, std::string>>> channelsByCycle;

        for (const auto& [systemId, _] : validReadings)
        {
            auto dcOpt = dclPackage.dataChannelList().from(systemId);
            if (dcOpt.has_value())
            {
                const auto& dc = dcOpt.value().get();
                auto updateCycleOpt = dc.property().dataChannelType().updateCycle();
                std::string cycleKey = updateCycleOpt.has_value() ? formatDouble(updateCycleOpt.value()) : "none";
                channelsByCycle[cycleKey].emplace_back(dc, systemId);
            }
        }

        std::cout << "Grouped channels into " << channelsByCycle.size() << " updateCycle groups:\n";
        for (const auto& [cycleKey, channels] : channelsByCycle)
        {
            std::string cycleStr = (cycleKey == "none") ? "None" : cycleKey + "s";
            std::cout << "  updateCycle=" << cycleStr << ": " << channels.size() << " channels\n";
        }
        std::cout << "\n";

        std::cout << "5. TimeSeriesData: Build TabularData for each group\n";
        std::cout << "------------------------------------------------------\n";

        std::vector<timeseries::TabularData> tabularDataList;

        for (const auto& [cycleKey, channelPairs] : channelsByCycle)
        {
            std::set<DateTimeOffset> allTimestamps;
            for (const auto& [_, systemId] : channelPairs)
            {
                for (const auto& reading : validReadings[systemId])
                {
                    allTimestamps.insert(reading.timestamp);
                }
            }

            std::vector<timeseries::DataChannelId> dataChannelIds;
            for (const auto& [dc, _] : channelPairs)
            {
                auto dcIdOpt = timeseries::DataChannelId::fromString(dc.dataChannelId().localId().toString());
                if (dcIdOpt.has_value())
                {
                    dataChannelIds.push_back(std::move(dcIdOpt.value()));
                }
            }

            std::vector<timeseries::TabularDataSet> dataSets;
            for (const auto& ts : allTimestamps)
            {
                std::vector<std::string> values;
                std::vector<std::string> qualities;

                for (const auto& [dc, systemId] : channelPairs)
                {
                    auto it = std::find_if(
                        validReadings[systemId].begin(), validReadings[systemId].end(), [&ts](const SensorReading& r) {
                            return r.timestamp == ts;
                        });

                    if (it != validReadings[systemId].end())
                    {
                        if (dc.property().format().type() == "Boolean")
                        {
                            values.push_back(it->value != 0.0 ? "True" : "False");
                        }
                        else
                        {
                            values.push_back(formatDouble(it->value));
                        }
                        qualities.push_back(it->quality);
                    }
                    else
                    {
                        values.push_back("");
                        qualities.push_back("Bad");
                    }
                }

                dataSets.push_back(timeseries::TabularDataSet{ ts, values, std::optional{ qualities } });
            }

            std::string cycleStr = (cycleKey == "none") ? "None" : cycleKey + "s";
            std::cout << "  TabularData: " << channelPairs.size() << " channels, " << dataSets.size()
                      << " rows (updateCycle=" << cycleStr << ")\n";

            tabularDataList.push_back(timeseries::TabularData{ dataChannelIds, dataSets });
        }
        std::cout << "\n";

        std::cout << "6. Package: Create TimeSeriesDataPackage with header\n";
        std::cout << "-------------------------------------------------------\n";

        std::vector<DateTimeOffset> allTimestampsList;
        for (const auto& [_, readings_vec] : validReadings)
        {
            for (const auto& reading : readings_vec)
            {
                allTimestampsList.push_back(reading.timestamp);
            }
        }

        if (allTimestampsList.empty())
        {
            std::cout << "No valid readings to build a TimeSeriesData package from.\n";
            return 0;
        }

        auto minTs = *std::min_element(allTimestampsList.begin(), allTimestampsList.end());
        auto maxTs = *std::max_element(allTimestampsList.begin(), allTimestampsList.end());

        auto shipIdOpt = ShipId::fromString("IMO1234567");
        if (!shipIdOpt.has_value())
        {
            throw std::runtime_error{ "Failed to parse shipId" };
        }
        auto shipId = shipIdOpt.value();

        const auto& dcHeader = dclPackage.package().header();
        auto configuration = timeseries::ConfigurationReference{ dcHeader.dataChannelListId().id(),
                                                                 dcHeader.dataChannelListId().timeStamp() };

        timeseries::Header tsHeader{ shipId,
                                     std::optional<timeseries::TimeSpan>{ timeseries::TimeSpan{ minTs, maxTs } },
                                     std::nullopt,
                                     std::nullopt,
                                     std::optional<std::string>{ "Vista SDK Sample" } };

        std::vector<timeseries::TimeSeriesData> timeSeriesDataVec;
        timeSeriesDataVec.push_back(
            timeseries::TimeSeriesData{ std::optional{ configuration }, std::optional{ tabularDataList } });

        timeseries::Package pkg{ std::optional{ tsHeader }, timeSeriesDataVec };
        auto tsPackage = timeseries::TimeSeriesDataPackage{ pkg };

        std::cout << "Created TimeSeriesDataPackage:\n";
        std::cout << "  Ship ID  : " << tsHeader.shipId().toString() << "\n";
        std::cout << "  TimeSpan : " << minTs.toString() << " to " << maxTs.toString() << "\n";
        std::cout << "  TabularData blocks: " << tabularDataList.size() << "\n";
        std::cout << "\n";

        std::cout << "7. Serialization: Convert to ISO19848 JSON format\n";
        std::cout << "----------------------------------------------------\n";

        auto jsonStr = jts::toJsonString(tsPackage, true);
        std::cout << "Serialized to JSON: " << jsonStr.size() << " bytes\n";
        std::cout << "Package:\n" << jsonStr << "\n";
        std::cout << "\n";

        std::cout << "8. Validation: Cross-check TimeSeriesData against DataChannelList\n";
        std::cout << "-----------------------------------------------------------------------\n";

        auto onData = [](const DateTimeOffset&,
                         const datachannel::DataChannel&,
                         const transport::Value&,
                         const std::optional<std::string>&) { return ValidateResult<>::ok(); };

        auto validation = timeSeriesDataVec.front().validate(dclPackage, onData, onData);

        std::cout << "  Is valid: " << std::boolalpha << bool{ validation } << "\n";
        if (!validation)
        {
            std::cout << "  Errors:\n";
            for (const auto& error : validation.errors())
            {
                std::cout << "    - " << error << "\n";
            }
        }
        std::cout << "\n";

        std::cout << "9. Validation failure: Out-of-range sensor value rejected\n";
        std::cout << "---------------------------------------------------------------\n";

        auto faultyChannelId = timeseries::DataChannelId::fromString(
                                   dclPackage.dataChannelList()["TEMP001"].dataChannelId().localId().toString())
                                   .value();

        auto faultyTabularData = timeseries::TabularData{
            std::vector<timeseries::DataChannelId>{ faultyChannelId },
            std::vector<timeseries::TabularDataSet>{ timeseries::TabularDataSet{
                baseTime, std::vector<std::string>{ "999.9" }, std::optional{ std::vector<std::string>{ "Good" } } } }
        };

        auto faultyTimeSeriesData =
            timeseries::TimeSeriesData{ std::optional{ configuration },
                                        std::optional{ std::vector<timeseries::TabularData>{ faultyTabularData } } };

        std::cout << "  Reporting TEMP001 = 999.9 degC (Restriction.MaxInclusive is 200)\n";

        auto faultyValidation = faultyTimeSeriesData.validate(dclPackage, onData, onData);

        std::cout << "  Is valid: " << std::boolalpha << bool{ faultyValidation } << "\n";
        if (!faultyValidation)
        {
            std::cout << "  Errors:\n";
            for (const auto& error : faultyValidation.errors())
            {
                std::cout << "    - " << error << "\n";
            }
        }
        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/transport/sensorsdataflow.rs
use std::collections::HashMap;

use vista_sdk::core::local_id_builder::OwnedLocalIdBuilder;
use vista_sdk::transport::datachannel::data_channel::{
    ConfigurationReference, DataChannel, DataChannelId as DclChannelId, DataChannelList,
    DataChannelListPackage, DataChannelType, Format, Header as DclHeader, Package as DclPackage,
    Property, Range, Restriction, Unit, VersionInformation,
};
use vista_sdk::transport::datachannel::data_channel_json;
use vista_sdk::transport::ship_id::ShipId;
use vista_sdk::transport::timeseries::data_channel_id::TsdChannelId;
use vista_sdk::transport::timeseries::time_series_data::{
    TabularData, TabularDataSet, TimeSeriesData, TimeSeriesDataPackage, TsdConfigRef, TsdHeader,
    TsdPackage, TsdTimeSpan,
};
use vista_sdk::transport::timeseries::time_series_data_json;
use vista_sdk::types::date_time_offset::DateTimeOffset;

struct SensorReading {
    system_id: String,
    value: f64,
    timestamp: DateTimeOffset,
    quality: String,
}

fn format_double(value: f64) -> String {
    format!("{}", value)
}

fn create_data_channel_list() -> DataChannelListPackage {
    let mut dcl = DataChannelList::new();

    let local_id1 = OwnedLocalIdBuilder::from_string(
        "/dnv-v2/vis-3-4a/411.1-1/C101.63/S206/meta/qty-temperature/cnt-cooling.air",
    )
    .expect("valid local id 1");

    let mut dc_id1 = DclChannelId::new(&local_id1);
    dc_id1.set_short_id("TEMP001");

    let mut restriction1 = Restriction::new();
    restriction1.set_fraction_digits(1);
    restriction1.set_max_inclusive(200.0);
    restriction1.set_min_inclusive(-50.0);

    let mut format1 = Format::new("Decimal");
    format1.set_restriction(&restriction1);

    let mut dct1 = DataChannelType::new("Inst");
    dct1.set_update_cycle(1.0);

    let range1 = Range::new(0.0, 150.0);
    let mut unit1 = Unit::new("°C");
    unit1.set_quantity_name("Temperature");

    let mut prop1 = Property::new(&dct1, &format1);
    prop1.set_range(&range1);
    prop1.set_unit(&unit1);
    prop1.set_quality_coding("OPC_QUALITY");
    prop1.set_name("Main Engine Air Cooler Temperature");

    dcl.add(&DataChannel::new(&dc_id1, &prop1));

    let local_id2 =
        OwnedLocalIdBuilder::from_string("/dnv-v2/vis-3-4a/511.15-1/E32/meta/qty-power")
            .expect("valid local id 2");

    let mut dc_id2 = DclChannelId::new(&local_id2);
    dc_id2.set_short_id("PWR001");

    let format2 = Format::new("Decimal");
    let mut dct2 = DataChannelType::new("Inst");
    dct2.set_update_cycle(1.0);
    let range2 = Range::new(0.0, 10000.0);
    let mut unit2 = Unit::new("kW");
    unit2.set_quantity_name("Power");

    let mut prop2 = Property::new(&dct2, &format2);
    prop2.set_range(&range2);
    prop2.set_unit(&unit2);
    prop2.set_name("Generator Power Output");

    dcl.add(&DataChannel::new(&dc_id2, &prop2));

    let local_id3 = OwnedLocalIdBuilder::from_string(
        "/dnv-v2/vis-3-4a/621.22i/S110/meta/cnt-diesel.oil/cmd-stop",
    )
    .expect("valid local id 3");

    let mut dc_id3 = DclChannelId::new(&local_id3);
    dc_id3.set_short_id("ALT001");

    let format3 = Format::new("Boolean");
    let dct3 = DataChannelType::new("Command");

    let mut prop3 = Property::new(&dct3, &format3);
    prop3.set_name("Diesel Oil Stop Command");

    dcl.add(&DataChannel::new(&dc_id3, &prop3));

    let ship_id = ShipId::from_string("IMO1234567").expect("valid ship id");
    let config_ref = ConfigurationReference::new(
        "DataChannelList-v1",
        DateTimeOffset::from_str("2026-08-28T00:00:00Z").unwrap(),
    );
    let mut version_info = VersionInformation::with_fields("dnv", "v2");
    version_info.set_reference_url("https://docs.vista.dnv.com");

    let mut header = DclHeader::new(&ship_id, &config_ref);
    header.set_version_information(&version_info);
    header.set_author("Vista SDK Sample");
    header.set_date_created(DateTimeOffset::from_str("2026-08-28T00:00:00Z").unwrap());
    let package = DclPackage::new(&header, &dcl);
    DataChannelListPackage::new(&package)
}

fn main() {
    println!("=== vista-sdk SensorDataFlow Sample ===\n");

    {
        println!("1. DataChannelList: Creating sample data channel list");
        println!("--------------------------------------------------------");

        let dcl_package = create_data_channel_list();

        let dcl_ref = dcl_package.data_channel_list().expect("data channel list");
        println!("Created DataChannelList with {} channels:", dcl_ref.len());

        for i in 0..dcl_ref.len() {
            let dc = dcl_ref.at(i).unwrap();
            let cycle = dc.property().data_channel_type().update_cycle();
            let cycle_str = cycle
                .map(|c| format!("{}s", format_double(c)))
                .unwrap_or_else(|| "None".to_string());
            let short_id = dc
                .channel_id()
                .short_id()
                .unwrap_or_else(|| "<no-short-id>".to_string());
            println!("  {} (updateCycle={})", short_id, cycle_str);
        }

        {
            println!();
            let json_str = data_channel_json::to_json(&dcl_package, true);
            println!("Serialized to JSON: {} bytes", json_str.len());
            println!("Package:\n{}", json_str);
        }
        println!();

        println!("2. Proprietary sensor data: Simulating sensor readings");
        println!("---------------------------------------------------------");

        let base_time = DateTimeOffset::from_str("2026-08-28T10:00:00Z").unwrap();
        let base_plus_5 = DateTimeOffset::from_str("2026-08-28T10:05:00Z").unwrap();

        let readings: Vec<SensorReading> = vec![
            SensorReading {
                system_id: "TEMP001".to_string(),
                value: 45.2,
                timestamp: base_time,
                quality: "Good".to_string(),
            },
            SensorReading {
                system_id: "TEMP001".to_string(),
                value: 46.1,
                timestamp: base_plus_5,
                quality: "Good".to_string(),
            },
            SensorReading {
                system_id: "PWR001".to_string(),
                value: 2500.0,
                timestamp: base_time,
                quality: "Good".to_string(),
            },
            SensorReading {
                system_id: "PWR001".to_string(),
                value: 2650.5,
                timestamp: base_plus_5,
                quality: "Good".to_string(),
            },
            SensorReading {
                system_id: "ALT001".to_string(),
                value: 1.0,
                timestamp: base_time,
                quality: "Good".to_string(),
            },
            SensorReading {
                system_id: "UNKNOWN".to_string(),
                value: 123.4,
                timestamp: base_time,
                quality: "Good".to_string(),
            },
        ];

        println!(
            "Created {} sensor readings from proprietary system",
            readings.len()
        );
        println!("Sample readings:");
        for r in &readings {
            println!(
                "  SystemID={}, Value={}, Timestamp={}",
                r.system_id, r.value, r.timestamp
            );
        }
        println!();

        println!("3. Validation: Filter readings against DataChannelList");
        println!("--------------------------------------------------------");

        let mut valid_readings: HashMap<String, Vec<&SensorReading>> = HashMap::new();

        for reading in &readings {
            let dcl_ref = dcl_package.data_channel_list().expect("data channel list");
            if dcl_ref.from_short_id(&reading.system_id).is_some() {
                valid_readings
                    .entry(reading.system_id.clone())
                    .or_default()
                    .push(reading);
                println!("  Accepted: {} = {}", reading.system_id, reading.value);
            } else {
                println!("  Rejected: {} (unknown channel)", reading.system_id);
            }
        }

        println!(
            "\nAccepted {} unique channels with valid data",
            valid_readings.len()
        );
        println!();

        println!("4. Grouping: Group channels by updateCycle (ISO19848 optimization)");
        println!("---------------------------------------------------------------------");

        let mut channels_by_cycle: HashMap<String, Vec<(String, String)>> = HashMap::new();

        for system_id in valid_readings.keys() {
            let dcl_ref = dcl_package.data_channel_list().expect("data channel list");
            if let Some(dc) = dcl_ref.from_short_id(system_id) {
                let update_cycle = dc.property().data_channel_type().update_cycle();
                let cycle_key = update_cycle
                    .map(|c| format_double(c))
                    .unwrap_or_else(|| "none".to_string());
                let local_id_str = dc.channel_id().local_id_string().unwrap_or_default();
                channels_by_cycle
                    .entry(cycle_key)
                    .or_default()
                    .push((local_id_str, system_id.clone()));
            }
        }

        println!(
            "Grouped channels into {} updateCycle groups:",
            channels_by_cycle.len()
        );
        for (cycle_key, channels) in &channels_by_cycle {
            let cycle_str = if cycle_key == "none" {
                "None".to_string()
            } else {
                format!("{}s", cycle_key)
            };
            println!("  updateCycle={}: {} channels", cycle_str, channels.len());
        }
        println!();

        println!("5. TimeSeriesData: Build TabularData for each group");
        println!("------------------------------------------------------");

        let mut tabular_data_list: Vec<TabularData> = Vec::new();

        for (cycle_key, channel_pairs) in &channels_by_cycle {
            let mut all_timestamps: Vec<DateTimeOffset> = Vec::new();
            for (_, system_id) in channel_pairs {
                if let Some(rs) = valid_readings.get(system_id) {
                    for r in rs {
                        if !all_timestamps
                            .iter()
                            .any(|t| t.ticks() == r.timestamp.ticks())
                        {
                            all_timestamps.push(r.timestamp);
                        }
                    }
                }
            }
            all_timestamps.sort_by_key(|t| t.ticks());

            let mut data_channel_ids: Vec<TsdChannelId> = Vec::new();
            for (local_id_str, _) in channel_pairs {
                if let Some(id) = TsdChannelId::from_string(local_id_str) {
                    data_channel_ids.push(id);
                }
            }

            let mut data_sets: Vec<TabularDataSet> = Vec::new();
            for &ts in &all_timestamps {
                let mut values: Vec<String> = Vec::new();
                let mut qualities: Vec<String> = Vec::new();

                for (_, system_id) in channel_pairs {
                    let dcl_ref = dcl_package.data_channel_list().expect("data channel list");
                    let format_type = dcl_ref
                        .from_short_id(system_id)
                        .map(|dc| dc.property().format().type_())
                        .unwrap_or_default();

                    if let Some(rs) = valid_readings.get(system_id) {
                        if let Some(r) = rs.iter().find(|r| r.timestamp.ticks() == ts.ticks()) {
                            if format_type == "Boolean" {
                                values.push(if r.value != 0.0 {
                                    "True".to_string()
                                } else {
                                    "False".to_string()
                                });
                            } else {
                                values.push(format_double(r.value));
                            }
                            qualities.push(r.quality.clone());
                        } else {
                            values.push(String::new());
                            qualities.push("Bad".to_string());
                        }
                    }
                }

                data_sets.push(TabularDataSet::new(
                    ts,
                    &values.iter().map(|s| s.as_str()).collect::<Vec<_>>(),
                    Some(&qualities.iter().map(|s| s.as_str()).collect::<Vec<_>>()),
                ));
            }

            let cycle_str = if cycle_key == "none" {
                "None".to_string()
            } else {
                format!("{}s", cycle_key)
            };
            println!(
                "  TabularData: {} channels, {} rows (updateCycle={})",
                channel_pairs.len(),
                data_sets.len(),
                cycle_str
            );

            tabular_data_list.push(TabularData::new(
                &data_channel_ids.iter().collect::<Vec<_>>(),
                &data_sets.iter().collect::<Vec<_>>(),
            ));
        }
        println!();

        println!("6. Package: Create TimeSeriesDataPackage with header");
        println!("-------------------------------------------------------");

        let mut all_timestamps_list: Vec<DateTimeOffset> = Vec::new();
        for rs in valid_readings.values() {
            for r in rs {
                all_timestamps_list.push(r.timestamp);
            }
        }

        if all_timestamps_list.is_empty() {
            println!("No valid readings to build a TimeSeriesData package from.");
            return;
        }

        let min_ts = *all_timestamps_list
            .iter()
            .min_by_key(|t| t.ticks())
            .unwrap();
        let max_ts = *all_timestamps_list
            .iter()
            .max_by_key(|t| t.ticks())
            .unwrap();

        let ship_id = ShipId::from_string("IMO1234567").expect("valid ship id");

        let pkg_ref = dcl_package.package().expect("package");
        let hdr_ref = pkg_ref.header().expect("header");
        let dcl_id = hdr_ref.data_channel_list_id();
        let configuration = TsdConfigRef::new(&dcl_id.id(), dcl_id.timestamp());

        let time_span = TsdTimeSpan::new(min_ts, max_ts).expect("valid time span");
        let mut ts_header = TsdHeader::new(&ship_id);
        ts_header.set_time_span(&time_span);
        ts_header.set_author("Vista SDK Sample");

        let mut ts_data = TimeSeriesData::new();
        ts_data.set_data_configuration(&configuration);
        ts_data.set_tabular_data(&tabular_data_list.iter().collect::<Vec<_>>());

        let package = TsdPackage::new(Some(&ts_header), &[&ts_data]);
        let tsd_package = TimeSeriesDataPackage::new(&package);

        println!("Created TimeSeriesDataPackage:");
        println!("  Ship ID  : IMO1234567");
        println!("  TimeSpan : {} to {}", min_ts, max_ts);
        println!("  TabularData blocks: {}", tabular_data_list.len());
        println!();

        println!("7. Serialization: Convert to ISO19848 JSON format");
        println!("----------------------------------------------------");

        let json_str = time_series_data_json::to_json(&tsd_package, true);
        println!("Serialized to JSON: {} bytes", json_str.len());
        println!("Package:\n{}", json_str);
        println!();

        println!("8. Validation: Cross-check TimeSeriesData against DataChannelList");
        println!("-----------------------------------------------------------------------");

        let validation = ts_data.validate(&dcl_package, &|_ts| Ok(()), &|_ts| Ok(()));

        println!("  Is valid: {}", validation.is_valid());
        if !validation.is_valid() {
            println!("  Errors:");
            for error in validation.errors() {
                println!("    - {}", error);
            }
        }
        println!();

        println!("9. Validation failure: Out-of-range sensor value rejected");
        println!("---------------------------------------------------------------");

        let dcl_ref = dcl_package.data_channel_list().expect("data channel list");
        let temp001_local_id = dcl_ref
            .from_short_id("TEMP001")
            .expect("TEMP001 channel")
            .channel_id()
            .local_id_string()
            .expect("local id");

        let faulty_channel_id =
            TsdChannelId::from_string(&temp001_local_id).expect("valid channel id");

        let faulty_ds = TabularDataSet::new(base_time, &["999.9"], Some(&["Good"]));
        let faulty_tabular = TabularData::new(&[&faulty_channel_id], &[&faulty_ds]);

        let mut faulty_ts_data = TimeSeriesData::new();
        faulty_ts_data.set_data_configuration(&configuration);
        faulty_ts_data.set_tabular_data(&[&faulty_tabular]);

        println!("  Reporting TEMP001 = 999.9 degC (Restriction.MaxInclusive is 200)");

        let faulty_validation = faulty_ts_data.validate(&dcl_package, &|_ts| Ok(()), &|_ts| Ok(()));

        println!("  Is valid: {}", faulty_validation.is_valid());
        if !faulty_validation.is_valid() {
            println!("  Errors:");
            for error in faulty_validation.errors() {
                println!("    - {}", error);
            }
        }
        println!();
    }
}
